File: webui/src/webui/mod/webui/cgibin/node.py
#! /usr/bin/env python

"""
usage: %(progname)s [args]
"""

import nstart

import os, sys, string, time, getopt, re
from pyclearsilver.log import *

import neo_cgi, neo_util, neo_cs

from pyclearsilver import CSPage
from pyclearsilver import odb

import MBPage
import db_webui

import roslib
import roslib.scriptutil

import webutil
import nodeutil



class MyPage(MBPage.MBPage):
  def setup(self, hdf):
    pass
    
  def display(self, hdf):
    node = hdf.getValue("Query.node", "")
    hdf.setValue("CGI.cur.node", node)
    webutil.set_tabs(hdf, ["status", "nodes"])
    webutil.grabTopics(hdf, ["/topics"])

    node_data = nodeutil.node_info(node)

    webutil.hdf_array(hdf, sorted(node_data["subscriptions"]), "CGI.cur.subscriptions")
    webutil.hdf_array(hdf, sorted(node_data["publications"]), "CGI.cur.publications")
    webutil.hdf_array(hdf, sorted(node_data["services"]), "CGI.cur.services")
    
    if "error" in node_data:
      hdf.setValue("CGI.cur.error", node_data["error"])
    
def run(context):
  return MyPage(context, pagename="node", nologin=False)

def main():
  context = CSPage.Context()
  run(context).start()

if __name__ == "__main__":
  main()

File: webui/src/webui/mod/webui/templates/node.cs
<?cs include:"ros_app_include_header.cs" ?>
<?cs include:"status_header.cs" ?>

<h3>Node: <?cs var:CGI.cur.node ?></h3>

<?cs if:CGI.cur.error ?>
  <span class="error_message">Uh oh!</span>  <?cs var:CGI.cur.error ?>
<?cs else ?>
  <h4>Publications</h4>
  <?cs each:_topic=CGI.cur.publications ?>
    <li><a href="topic.py?topic=<?cs var:_topic ?>"><?cs var:_topic ?></a></li>  
  <?cs /each ?>

  <h4>Subscriptions</h4>
  <?cs each:_topic=CGI.cur.subscriptions ?>
    <li><a href="topic.py?topic=<?cs var:_topic ?>"><?cs var:_topic ?></a></li>  
  <?cs /each ?>

  <h4>Services</h4>
  <?cs each:_service=CGI.cur.services ?>
    <li><a href="service.py?service=<?cs var:_service ?>"><?cs var:_service ?></a></li>  
  <?cs /each ?>
<?cs /if ?>

<p>
  <a href="nodes.py">&laquo; All nodes</a>
</p>

<?cs include:"ros_app_include_footer.cs" ?>
